tty/spike-tty/spike-tty.c

```diff
@@ -1,191 +1,153 @@
 /*
  * Phoenix-RTOS
  *
- * Operating system kernel
- *
  * Spike tty (HTIF console)
  *
- * Copyright 2020 Phoenix Systems
- * Author: Pawel Pisarczyk, Julia Kosowska
+ * Copyright 2020, 2021 Phoenix Systems
+ * Author: Pawel Pisarczyk, Julia Kosowska, Lukasz Kosinski
  *
  * %LICENSE%
  */
 
 #include <errno.h>
-#include <poll.h>
 #include <stdint.h>
 #include <stdio.h>
-#include <stdlib.h>
-#include <string.h>
-#include <sys/threads.h>
-#include <sys/interrupt.h>
-#include <sys/io.h>
-#include <sys/file.h>
 #include <unistd.h>
+
+#include <sys/file.h>
+#include <sys/io.h>
 #include <sys/msg.h>
-#include <sys/mman.h>
+#include <sys/threads.h>
+#include <sys/types.h>
 
 #include <libtty.h>
 
 
 typedef struct {
-//	handle_t mutex;
+	int active;
 	oid_t oid;
 	libtty_common_t tty;
+	char stack[1024] __attribute__((aligned(8)));
 } spiketty_t;
 
 
-static spiketty_t *spikettys[1];
-
-static spiketty_t spiketty = {0};
-
-static void set_baudrate(void *arg, speed_t baud)
-{
-	/* TODO */
-}
-
-
-static void set_cflag(void *arg, tcflag_t *cflag)
-{
-	/* TODO */
-}
-
-
-static void signal_txready(void *arg)
-{
-	spiketty_t *spiketty = (spiketty_t *)arg;
-
-	while (libtty_txready(&spiketty->tty))
-		sbi_putchar(libtty_getchar(&spiketty->tty, NULL));
-}
+static struct {
+	spiketty_t spikettys[1];
+	char stack[2048] __attribute__((aligned(8)));
+} spiketty_common;
 
 
-static int spiketty_write(uint8_t d, size_t len, char *buff, int mode)
+static spiketty_t *spiketty_get(oid_t *oid)
 {
 	spiketty_t *spiketty;
+	unsigned int i;
 
-	if (d >= sizeof(spikettys) / sizeof(spiketty_t *))
-		return -EINVAL;
-
-	if ((spiketty = spikettys[d]) == NULL)
-		return -ENOENT;
+	for (i = 0; i < sizeof(spiketty_common.spikettys) / sizeof(spiketty_common.spikettys[0]); i++) {
+		spiketty = spiketty_common.spikettys + i;
 
-	if (!len)
-		return 0;
+		if (spiketty->active && (spiketty->oid.port == oid->port) && (spiketty->oid.id == oid->id))
+			return spiketty;
+	}
 
-	return libtty_write(&spiketty->tty, buff, len, mode);
+	return NULL;
 }
 
 
-static int spiketty_read(uint8_t d, size_t len, char *buff, int mode)
+static void set_baudrate(void *arg, speed_t baud)
 {
-	spiketty_t *spiketty;
-
-	if (d >= sizeof(spikettys) / sizeof(spiketty_t *))
-		return -EINVAL;
-
-	if ((spiketty = spikettys[d]) == NULL)
-		return -ENOENT;
-
-	return libtty_read(&spiketty->tty, buff, len, mode);
+	/* TODO */
 }
 
 
-static int spiketty_status(uint8_t d)
+static void set_cflag(void *arg, tcflag_t *cflag)
 {
-	spiketty_t *spiketty;
-
-	if (d >= sizeof(spikettys) / sizeof(spiketty_t *))
-		return POLLNVAL;
-
-	if ((spiketty = spikettys[d]) == NULL)
-		return POLLNVAL;
-
-	return libtty_poll_status(&spiketty->tty);
+	/* TODO */
 }
 
 
-uint8_t spiketty_get(oid_t *oid)
+static void signal_txready(void *arg)
```

Comment:
I know that's outside of the scope of this PR, but this is a one weird function... Instead of signaling txready we push data from libtty (without even checking if there's any data available). Why?

Reply:
`spike-tty` driver doesn't register any interrupts. We don't receive any TX ready signal. We're constantly ready to write new data with `sbi_putchar()` therefore we can do in `signal_txready` which is a callback called by `libtty` whenever new data is available in the TX fifo.
Also we do have a check for available data - `libtty_txready()` returns `true` as long as there is data in TX fifo to be written.